### PR TITLE
Add MAML

### PR DESCRIPTION
**Pull request by Yash-10**
Currently, the `task_accuracy` and `task_loss` functions assume classification. Probably better to keep them separate.

---

Maybe a topic for a separate issue: Would be helpful to simultaneously work on developing methods for creating and sampling tasks that do not use `learn2learn`.

Any ideas @veds12 ?

### CODE REVIEW

**Comment by veds12** [comment]
@Yash-10 I am thinking about a general initial skeleton for the library - trainers, samplers etc. Once we have an initial plan, we can restructure this PR according to that. 

Writing a general task sampler now would be definitely useful as a starting point so you can work on it now and we can go over it together once done.

File: src/max/maml.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
+import numpy as np
+from functools import partial
+
+import jax
+import optax
+import jax.numpy as jnp
+from flax import nn, optim
+
+@jax.jit
+def task_accuracy(logits, labels):
```
Comment (veds12) [inline src/max/maml.py]: Shift this to `max/metrics.py`

File: src/max/maml.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    """
+    model_grad = jax.grad(task_loss)(model, batch)
+    # Create a new optimizer for the given `model`. This is analogous to creating a learner for each task using `.clone()`.
+    inner_opt = inner_optimizer_def.create(model)
+    for _ in range(fas):  # Do fast adaptation for `fas` number of times.
+        inner_opt = inner_opt.apply_gradient(model_grad)  # Analogous to `learner.adapt` step from `learn2learn`.
+    return inner_opt.target  # return the updated model which is stored as an attribute of the updated optimizer.
+
+@jax.jit
+def train_step(optimizer, batch):
```
Comment (veds12) [inline src/max/maml.py]: Shift to 'max/opti_trainer.py`

Comment (veds12) [reply src/max/maml.py]: Can you also try to make `maml_loss` more algorithm agnostic?

File: src/max/maml.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+import jax.numpy as jnp
+from flax import nn, optim
+
+@jax.jit
+def task_accuracy(logits, labels):
+    """Calculates classification accuracy for a task."""
+    return jnp.mean(jnp.argmax(logits, -1) == labels)
+
+@jax.jit
+def task_loss(model, batch):
```
Comment (veds12) [inline src/max/maml.py]: Shift to `max/loss.py`

**Comment by veds12** [comment]
@Yash-10 I have added some comments to the code. Could you look into those? Basically trying to make the entire thing as modular as possible

File: src/max/maml.py (hunk 2 above)
Comment (Yash-10) [reply src/max/maml.py]: > Can you also try to make maml_loss more algorithm agnostic?

Great idea! Changed `maml_loss` to `loss` and instead allowed passing a generalized `fit_task` function to `train_step`. So `maml_fit_task` from `maml.py` could be passed as an argument, for example.

**Comment by Yash-10** [comment]
Seems a bit more modular.

I believe `fit_task` could in principle might have other model-specific positional/keyword arguments for which *args/**kwargs might be a good option. Need to see if this could be done in a straightforward manner for jax jitted functions.

**Comment by veds12** [comment]
That's right. Not just `fit_task` but all the other functions which can be reused across different algorithms. We'll need to add *args and **kwargs as we scale.

**Comment by veds12** [comment]
@Yash-10 can you squash all your commits into 1?

**Comment by Yash-10** [comment]
Done!

Should we add *args/**kwargs here or tackle this in another PR in the future? If its kept in the future, I might remove the TODOs in the code and instead create a new issue.

**Comment by veds12** [comment]
Thanks. We can add *args and **kwargs later when we add other algorithms. 
Also, can you mark the PR it ready to merge (instead of Draft). I won't be able to merge otherwise

**Comment by Yash-10** [comment]
Done!
